var/spack/repos/builtin/packages/py-pennylane-lightning/package.py

```diff
@@ -7,20 +7,124 @@
+    variant("verbose", default=False, description="Build with full verbosity")
+
+    variant("cpptests", default=False, description="Build CPP tests")
+    variant("cppbenchmarks", default=False, description="Build CPP benchmark examples")
+
+    variant(
+        "build_type",
+        default="Release",
+        description="CMake build type",
+        values=("Debug", "Release", "RelWithDebInfo", "MinSizeRel"),
+    )
+
+    extends("python")
+
+    # hard dependencies
+    depends_on("cmake@3.21:3.24,3.25.2:", type="build")
+    depends_on("ninja", type=("run", "build"))
+
+    # variant defined dependencies
+    depends_on("blas", when="+blas")
+    depends_on("kokkos@3.7.00", when="+kokkos")
+    depends_on("kokkos-kernels@3.7.00", when="+kokkos")
+    depends_on("llvm-openmp", when="+openmp %apple-clang")
 
     depends_on("python@3.8:", type=("build", "run"))
     depends_on("py-setuptools", type="build")
-    depends_on("cmake@3.16:", type="build")
-
-    depends_on("py-ninja", type=("build", "run"))
     depends_on("py-numpy", type=("build", "run"))
-    # depends_on("py-pennylane@0.19:", type=("build", "run"))  # circular dependency
+    depends_on("py-pybind11", type=("build"))
+    depends_on("py-pip", type="build")
+    depends_on("py-wheel", type="build")
+
+
+class CMakeBuilder(spack.build_systems.cmake.CMakeBuilder):
+    build_directory = "build"
+
+    def cmake_args(self):
+        """
+        Here we specify all variant options that can be dynamicaly specified at build time
+        """
+        args = [
+            self.define_from_variant("CMAKE_BUILD_TYPE", "build_type"),
+            self.define_from_variant("ENABLE_OPENMP", "openmp"),
+            self.define_from_variant("ENABLE_NATIVE", "native"),
+            self.define_from_variant("ENABLE_BLAS", "blas"),
+            self.define_from_variant("CMAKE_VERBOSE_MAKEFILE:BOOL", "verbose"),
+            self.define_from_variant("BUILD_TESTS", "cpptests"),
+            self.define_from_variant("BUILD_BENCHMARKS", "cppbenchmarks"),
+            self.define_from_variant("ENABLE_GATE_DISPATCHER", "dispatcher"),
+        ]
+
+        if self.spec.variants["kokkos"].value:
+            args += [
+                "-DENABLE_KOKKOS=ON",
+                f"-DKokkos_Core_DIR={self.spec['kokkos'].home}",
+                f"-DKokkos_Kernels_DIR={self.spec['kokkos-kernels'].home}",
+            ]
+        else:
+            args += ["-DENABLE_KOKKOS=OFF"]
+
+        return args
+
+    def build(self, pkg, spec, prefix):
+        super().build(pkg, spec, prefix)
+        cm_args = ";".join(
+            [
+                s[2:]
+                for s in self.cmake_args()
+                if s[2:] not in ["BUILD_TESTS:BOOL=ON", "BUILD_BENCHMARKS:BOOL=ON"]
+            ]
+        )
+        args = ["-i", f"--define={cm_args}"]
+        build_ext = Executable(f"{self.spec['python'].command.path}" + " setup.py build_ext")
+        build_ext(*args)
```

Comment:
Still curious to see if @pradyunsg knows of a way to pass this option via pip instead of invoking setup.py.

Reply:
Hi @adamjstewart is there an update here? We explored the use of the Python `build` package as an alternative, but do not think this is an improvement over the `build_ext` we have above, due to the increase in verbosity: as an example 
```bash
python -m build --wheel --config-setting=--build-option='build_ext --define="CMAKE_CXX_COMPILER=clang-14;ENABLE_CLANG_TIDY=on"'
```
 would be required, instead of 
```bash
python setup.py build_ext --define="CMAKE_CXX_COMPILER=clang-14;ENABLE_CLANG_TIDY=on"
```
Since this isn't an issue for anything pre Python 3.12, which we do not aim to support until sometime in 2024, I think we can keep the above for now, as many other packages are doing likewise, and aim to identify a cleaner solution that is more maintainable long-term.

In addition, we have updated the package file to the newest release of both PennyLane and PennyLane-Lightning. Feel free to let us know if this meets the cut.

Reply:
I'm fine with keeping this as is for now, we can improve it at a later date as needed.

Reply:
Sounds good, thanks!